pkg/hashcircler/hash_circler.go

```diff
@@ -0,0 +1,253 @@
+	if hashFunc == nil {
+		hashFunc = fnvHashFunc
+	}
+
+	if len(keys) == 0 {
+		return nil, fmt.Errorf("empty keys")
+	}
+
+	keyArr := make([]*keyStatus, len(keys))
+
+	for i, key := range keys {
+		keyArr[i] = &keyStatus{
+			key:    key,
+			enable: true,
+		}
+	}
+
+	hc := &preSetHashCircler{
+		keyArr:        keyArr,
+		hashFunc:      hashFunc,
+		validKeyCount: len(keyArr),
+	}
+
+	vNodes := hc.generateVirtualNodes(keys, 32)
+	hc.vNodes = vNodes
+
+	return hc, nil
+}
+
+func (h *preSetHashCircler) Enable(key string) error {
+	h.Lock()
+	defer h.Unlock()
+
+	found := false
+
+	for _, k := range h.keyArr {
+		if k.key == key {
+			found = true
+			if k.enable {
+				return nil
+			}
+
+			k.enable = true
+			break
+		}
+	}
+
+	if !found {
+		return ErrKeyNotPresent
+	}
+
+	h.validKeyCount++
+	return nil
+}
+
+func (h *preSetHashCircler) Hash(input string) (key string, err error) {
+	h.RLock()
+	defer h.RUnlock()
+
+	if h.validKeyCount == 0 {
+		return "", ErrKeyNotPresent
+	}
+
+	hashIndex := h.hashFunc(input)
+	rangeIndex := sort.Search(len(h.vNodes), func(i int) bool {
+		return hashIndex <= h.vNodes[i].end
+	})
+
+	currIndex := rangeIndex
+	for {
+		keyIndex := h.vNodes[rangeIndex].keyIndex
+		if h.keyArr[keyIndex].enable {
+			return h.keyArr[keyIndex].key, nil
+		}
+
+		rangeIndex = h.circlePrvIndex(rangeIndex, len(h.vNodes))
+		if rangeIndex == currIndex {
+			break
+		}
+	}
+
+	return "", ErrKeyNotPresent
+}
+
+func (h *preSetHashCircler) Disable(key string) {
+	h.Lock()
+	defer h.Unlock()
+
+	found := false
+
+	for _, k := range h.keyArr {
+		if k.key == key {
+			found = true
+			if !k.enable {
+				return
+			}
+
+			k.enable = false
+			break
+		}
+	}
+
+	if !found {
+		return
+	}
+
+	h.validKeyCount--
+	return
+}
+
+func (h *preSetHashCircler) circlePrvIndex(index int, arrCount int) int {
+	if index > 0 {
+		return index - 1
+	}
+
+	return arrCount - 1
+}
+
+func (h *preSetHashCircler) circleNextIndex(index int, arrCount int) int {
+	if index < arrCount-1 {
+		return index + 1
+	}
+
+	return 0
+}
+
+func (h *preSetHashCircler) generateVirtualNodes(keys []string, virtualNodesPerKey int) []*virtualNode {
+	dup := map[uint64]struct{}{}
+	vNodesArr := []*virtualNode{}
+	for i := range keys {
+		indexArr := h.randomVirtualNodeKeys(dup, virtualNodesPerKey)
+		for _, index := range indexArr {
+			vNodesArr = append(vNodesArr, &virtualNode{
+				start:    index,
+				keyIndex: i,
+			})
+		}
+	}
+
+	sort.Slice(vNodesArr, func(i, j int) bool {
+		return vNodesArr[i].start < vNodesArr[j].start
+	})
+
+	for i := 0; i < len(vNodesArr)-1; i++ {
+		vNodesArr[i].end = vNodesArr[i+1].start - 1
+	}
+
+	vNodesArr[0].start = 0
+	vNodesArr[len(vNodesArr)-1].end = math.MaxUint64
+
+	return vNodesArr
+}
+
+func (h *preSetHashCircler) randomVirtualNodeKeys(dup map[uint64]struct{}, count int) []uint64 {
+	index := 0
+	res := make([]uint64, count)
+	for {
+		if index >= count {
+			break
+		}
+
+		v := rand.Uint64()
```

Comment:
It's not a good solution to use `rand.Unit64()` with a `rand.Seed(time.Now().UnixNano())` directly to create virtual nodes, it may cause 2 problems:
1. it gets different results in different nodes and each execution, this is unexpected
2. it cannot guarantee that the virtual nodes are evenly distributed

Reply:
Now [0, math.MaxUint64] is spilt into pieces, and range pieces will be evenly distributed by keys.

Reply:
Once keys  are determined, the owners which the divided range belong to are determined again.